### PR TITLE
feat: add HashMap type and Int kind

### PR DESCRIPTION
**Pull request by rvagg**
HAMT logic is in https://github.com/rvagg/iamap (not using the term "HAMT" anywhere here, how about we call it a "HashMap"). I'm working on a WIP modified version @ https://github.com/rvagg/iamap/pull/1, hence the git dependency.

Will comment inline.

### CODE REVIEW

File: src/kinds.js [hunk 1]
```diff
@@ -42,6 +42,17 @@ class StringKind extends Type {
 
 exports.StringKind = StringKind
 
+class IntKind extends Type {
```
Comment (rvagg) [inline src/kinds.js]: I wanted to test returning an integer so figured this was appropriate

File: src/maps/hashmap.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+const IAMap = require('iamap')
+const murmurhash3 = require('murmurhash3js-revisited')
+const Type = require('../base')
+const Block = require('@ipld/block')
+
+const _type = 'IPLD/Experimental/HashMap/0'
```
Comment (rvagg) [inline src/maps/hashmap.js:6]: I still can't _not_ version these things .. even with "Experimental"

File: src/maps/hashmap.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    try {
+      let result = await map.get(key)
+      if (!path.length) {
+        return { result }
+      }
+      throw new Error('Traversing into values is not yet supported')
+      // what do? could be an arbitrary kind
+      // return { call: { info: { method: 'get', args: path.join('/') }, target: result, proxy: true } }
```
Comment (rvagg) [inline src/maps/hashmap.js]: I don't think this is supported yet but I'd like to be able to return an arbitrary thing and be able to navigate into it, e.g. the map could store objects inline and you should be able to skip into those objects via remaining parts of `path`, see tests for examples of what I think you should be able to do.

File: src/maps/hashmap.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      throw new Error('HashMap type needs a string to look up')
+    }
+
+    // TODO: this should be a validation of this.node against the HAMT root block schema
+    if (!IAMap.isRootSerializable(this.node) || !IAMap.isSerializable(this.node)) {
+      throw new Error('Block does not contain expected HashMap root data')
+    }
+
+    // provide a store to IAMap that won't do any retrieval but will intercept load() calls
+    // so we can pass that up for continuation
```
Comment (rvagg) [inline src/maps/hashmap.js]: IAMap assumes it has a backing store with `save()`, `load()` and an `isEqual(id1, id2)` as it's agnostic to storage system and what an "id" is. Mutation and fetch operations work on the storage system as it needs to save new blocks or load existing ones. So in the get() situation we're only going to load blocks and loading here means passing CIDs and continuations, hence this `fauxStorage` thing which does the intercepting.

File: src/maps/hashmap.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      let call = { info, target }
+      return { call }
+    }
+  }
+}
+
+HashMap._type = _type
+
+// register an x86 32-bit murmurhash as available for use and make it the default (in create() below)
+// this is probably not what we want to end up doing
```
Comment (rvagg) [inline src/maps/hashmap.js]: here's something for you to ponder and connects to a discussion in https://github.com/ipld/specs/pull/109 about HAMT properties—you should be able to configure what hash function you're using, but we can't support every conceivable hash that might be usable so I've built in a `registerHasher()` mechanism to make it fully pluggable.
But having that much flexibility gets in the way of the transparent navigation here, a user shouldn't have to register basic hashes that are commonly in use.

So I'm thinking that re HAMT spec:

1. hash function should be configurable
2. HAMT (or "HashMap") implementaitons should ship with a good fast hash and a good cryptographic hash by default, maybe murmur-x64-128 and sha2-256
3. if a HAMT you want to navigate through uses some alternate hash, there should be some way to register that through this type system

that registering process isn't likely to be unique to the HAMT either, encryption is going to have all sorts of flexibility but I can imagine other collections having configurability that we may want to leave pluggable to allow for the user to make it work best for their use-case which may be considerably different to what we design for.

Here I'm just registering murmur-x86-32 for simplicity.

File: test/map.hashmap.spec.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  map = await map.set('foo', 'bar')
+  map = await map.set('bar', 100)
+  map = await map.set('baz', { wut: { is: 'this' } })
+  map = await map.set('boom', externalCid)
+
+  let root = map.id
+
+  same((await get(store, root, 'foo')).toString(), 'bar')
+  same((await get(store, root, 'bar')).toInt(), 100)
+  /* TODO: we should be able to navigate into child properties of values
```
Comment (rvagg) [inline test/map.hashmap.spec.js]: here are the parts where it seems that the path should allow us to navigate into objects stored inline in map values and transparently across links to other blocks where the value is a CID

File: src/maps/hashmap.js (hunk 2 above)
Comment (mikeal) [reply src/maps/hashmap.js:6]: Ya, you’re right, I’m going to go in and update the rest to do the same.

File: src/maps/hashmap.js (hunk 4 above)
Comment (mikeal) [reply src/maps/hashmap.js]: Ah, clever. Once we have the write side of this figured out we should be able to migrate the whole implementation over and no longer need this but this is a nice hack in the interim.

File: src/maps/hashmap.js (hunk 3 above)
Comment (mikeal) [reply src/maps/hashmap.js]: you should already be able to do the thing you’ve commented out. this works for `get` because the path is an arg. any target you return that isn’t a string is treated as the target directly, this should work. it end up not working in the non-`get` case if you need (target + path) because other methods won’t resolve the path as an arg and you only get one thing you’re allowed to return as the target.

File: src/maps/hashmap.js (hunk 5 above)
Comment (mikeal) [reply src/maps/hashmap.js]: We have a similar issue in the `unixfsv2` spec (once I update it) in that we want to say “This property can be any Binary Type” but we obviously can’t ensure compatibility if we don’t have a list of required types to be supported.

The solution I’ve been kicking around in my head is, each version of spec has a list of required multi-block types. If we implement better multi-block types in the future then we put out a new version of the spec noting the new type requirements.

We should do something similar here, the spec should require support for `murmur-x64-128` but the implementation can leave the door open to additional algorithms in future versions.

File: test/map.hashmap.spec.js (hunk 6 above)
Comment (mikeal) [reply test/map.hashmap.spec.js]: I’ll implement a `path` method that does just that. I’d like the `get` method to stay close to the underlying type method but we should have a recursive version for full path traversal.

The funny thing is, we already do full path traversal *inside* the implementation when you return a path as a target :)

Comment (mikeal) [reply test/map.hashmap.spec.js]: Actually, forget everything I’ve said up until now :)

https://github.com/ipld/js-types/blob/master/src/kinds.js#L14

We should already do this ;) The type implementation shouldn’t return a `remaining` path, it should return a `get` for the remaining path if it can’t fulfill the entire path locally. I forgot that I made this upgrade in the wee hours.

It makes sense, this reduces the work load quite a bit. If someone wants to do a more incremental path resolution they can just break up the path themselves and resolve each part one at a time. The default behavior should be the most efficient, which is to pass along the entire path and resolve it entirely.

Comment (mikeal) [reply test/map.hashmap.spec.js]: It’s probably time to document this since I can’t even keep it all straight myself.

File: src/maps/hashmap.js (hunk 5 above)
Comment (rvagg) [reply src/maps/hashmap.js]: I'd just like us to keep in mind JS API considerations for how to register / side-load alternative hash algorithms for users that may want to do that. It'd either be similar to the `Lookup` mechanism you have or just some global you could hook in to if you absolutely need to. Right now it's `require('iamap').registerHasher(...)` and it'd be registered globally, but it'd be good to have that integrated as some kind of side loading mechanism.

File: test/map.hashmap.spec.js (hunk 6 above)
Comment (rvagg) [reply test/map.hashmap.spec.js]: I had `args{}` wrong, got it working now and it's pretty darn sweet

File: src/kinds.js (hunk 1 above)
Comment (mikeal) [reply src/kinds.js]: Yup, I had been meaning to do this.

File: src/maps/hashmap.js (hunk 5 above)
Comment (mikeal) [reply src/maps/hashmap.js]: If you make it configurable on the class instead of the module then it would match up nicely with the existing Lookup mechanism since you pass in a specific class.

Comment (rvagg) [reply src/maps/hashmap.js]: The problem is: how do we expose the class? But I guess the answer to that depends on the what the writer interface will end up looking like. Currently that's these hacky `create()` factories but we need something nicer. Something that extends a `NodeBuilder` like in Go would make sense. It'd be really nice to have a base that does all of the saving of blocks and even build in transactional semantics to make mutations cheaper.
Something to defer for now. Perhaps revisit when we have more complex multiblock data structures with weird many-block mutation semantics.

**Comment by rvagg** [comment]
So, latest update to the `get` internals have removed the hacks. The assumption I'm working on now is that many of these data structure algorithms are going to be sufficiently complex that we're not going to want to bloat this repo with them. They'll also be easier to document and test (and they'll need heavy tests) as stand-alone packages and repos. So "integrating" is just a matter of interfacing.

That was what I was doing with `IAMap` -- boil down the essence of the multi-block asynchronous immutable data structure, detached from the storage platform and even specifics about what "blocks" mean other than simple JS objects that get serialised, deserialsed and have identifiers attached to them.

The question has been: what does the interface to such an abstraction look like and that depends on how we consume it here (and perhaps other places--maybe the writer will live in a different place? maybe there'll be reason to consume it in alternate forms). And with `get`, that's becoming more clear: the interface is "traverse a single block, with some continuation/context data".

The interface is now all static methods that operate on blocks, with the core single-block traversal operation being exposed as: [`IAMap.traverse(rootBlock, currentBlock, depth, key, isEqual)`](https://github.com/rvagg/iamap/tree/modified-serialization#IAMap__traverse) - where you traverse a single `currentBlock` with `rootBlock` as the context and `depth` as the distance from the root, `key` is the value you're looking up and `isEqual` is just `cid1.equals(cid2)` so IAMap can be identifier agnostic. It returns a `{ value, nextId }` pair where `nextId` is non-null where a traversal needs to span to a further child `depth+1` away from the root. This is all synchronous (may need to be async for browser crypto) and gives us a way to translate to `{ result }` or `{ call }`.

I have one more major change to the internals of the HAMT algorithm before it's in a state where I'd be willing to propose a spec based on it (which will be pretty close to the existing Go HAMT and the existing spec proposal). But I want to move on to a sorted data structure to try and extend this way of interfacing with the multi-block algorithms. Maybe the write strategy will become clearer when we have a diversity of algorithms.

**Comment by mikeal** [comment]
> This is all synchronous (may need to be async for browser crypto) and gives us a way to translate to { result } or { call }.

The system always uses `await` so if you need async you’ve got it :)

**Comment by mikeal** [comment]
Just shipped a bunch of changes to the internals.

* `Type` has been renamed to `Node`
* `this.node` inside subclasses has been renamed to `this.data`
* Lookup.register() can take a single argument for the class assuming there is a class attribute for `_type`.

I took your changes to add support for the IntKind as well. 

As far as “how should these be exported/consumed in the future” I think it will probably look something like:

```javascript
const iq = require(@ipld/iq)
const IAMap = require(‘@ipld/iamap’)

iq.config.lookup.register(IAMap) // we can maybe shorten this to `iq.register()`
```

**Comment by rvagg** [comment]
`iq.config.lookup.register()` is a global static then, there's no `new Lookup()` like now? I think I liked the way that it's instance-local rather than global, that's not a critical difference though.

So you can register a data structure, but let's think about how you might register metadata that is specific to some data structure so each one doesn't have its own custom way.

So for a map type I currently have to `IAMap.registerHasher(codec, hashBytes, hasherFn)` if I want to use a different hash function than one that's available. I'd like to be able to do that on `iq` somewhere. Can we make `Lookup` more flexible such that registered components can use it as a bucket?

Something like `iq.config.lookup.set('iamap.hasher.murmur3-32', { hashBytes, hasherFn })`. Then, since the `Type` base has a copy of the `Lookup` instance it could fetch these values, `this.lookup.get('iamap.hasher.murmur3-32')`.

The only problem I see with this is that it makes `iq.config.lookup` do what it sounds like `iq.config` should be doing. What else do you imagine `iq.config` being used for?

**Comment by rvagg** [comment]
rebased

I got IAMap all tidied up and released with the algorithm updates I was intending, so the git dependency is gone now.

It's a bit neater now, with root blocks looking like:

```
{
  hashAlg: string
  bitWidth: number
  bucketSize: number
  map: number
  data: Array
}
```

and intermediate and leaf blocks looking like:

```
{
  map: number
  data: Array
}
```

Next step is to ipld/spec it.

**Comment by mikeal** [comment]
> iq.config.lookup.register() is a global static then, there's no new Lookup() like now? I think I liked the way that it's instance-local rather than global, that's not a critical difference though.

That API is just for `iq`, this library `@ipld/types` still works the same and requires an explicit `lookup` object in the config.

In `iq`, there is a global default you can mess with but you can also get a new instance with whatever config overrides you want. You can see all the relevant code here https://github.com/ipld/js-iq/blob/master/index.js#L155

**Comment by rvagg** [comment]
Implemented `keys()`

I'm also trying out a new form of block-by-block traversal and state management between blocks. In the process I have `values()` and `entries()` (emits `{key,value}`) in the same form too, so we could add them here pretty easily too.

I had to use `{ local: true }` to prevent intermediate blocks from showing up in index.js/`keys()`, because of `leaf: !info`, which gets really messy because these blocks have their own `data` property so `for (let key of line.result.data)` actually tries to do stuff on them. Do you think it's appropriate to use `local` in this way or is this a bug?

**Comment by mikeal** [comment]
@rvagg sounds like a bug, `local` is suppose to be set by the type system and not by the implementor.

I *think* I found the bug and fixed https://github.com/ipld/js-types/commit/8e3bf2de36411a77c6d7b36ffbb2dfc3447a7b97

Try backing out your setting of `local` and see if it works now.

**Comment by mikeal** [comment]
This surfaces the need for much more extensive tests before we consider this ready for other people to use.

Let’s figure out a good way to include the HAMT implementation and its tests in this repo for testing purposes.

In fact, I’m not opposed to loading this repo up with example implementations and their tests. The way this is written I’m much less confident in the coverage reports because lines are easily covered but the implied variances in behavior isn’t without a lot of type implementations.

**Comment by rvagg** [comment]
latest changes fixed it, I no longer need `local:true`

**Comment by mikeal** [comment]
@rvagg perfect, thanks!
